harness/determined/pytorch/trainer.py

```diff
@@ -0,0 +1,138 @@
+        aggregation_frequency: Optional[int] = 1,
+        checkpoint_policy: Optional[str] = "best",
+        smaller_is_better: Optional[bool] = True,
+    ):
+
+        if self.local_training:
+            assert (max_epochs is None) ^ (max_batches is None), \
+                "Either max_batches or max_epochs must be defined in local training mode"
+            if max_epochs:
+                assert records_per_epoch, "records_per_epoch must be specified if training with epoch units"
+        else:
+            if max_batches or max_epochs:
+                logging.warning("max_batches and max_epochs is ignored in non-local-training mode. "
+                                "Please configure the searcher length instead.")
+
+        # Set context and training variables
+        self.context._aggregation_frequency = aggregation_frequency
+        self.context._average_aggregated_gradients = average_aggregated_gradients
+
+        max_length = None
+        if max_batches:
+            max_length = Batch(max_batches)
+        elif max_epochs:
+            max_length = Epoch(max_epochs)
+
+        searcher_metric_name = self.cluster_info and self.cluster_info.trial._config["searcher"]["metric"] or None
+        records_per_epoch = records_per_epoch or self.cluster_info and self.cluster_info.trial._config["records_per_epoch"]
+        steps_completed = self.cluster_info and self.cluster_info.trial._steps_completed
+
+        self.trial_controller = PyTorchTrialController(
+            trial_inst=self.trial,
+            context=self.context,
+            max_length=max_length,
+            min_checkpoint_period=min_checkpoint_period,
+            min_validation_period=min_validation_period,
+            average_training_metrics=average_training_metrics,
+            checkpoint_policy=checkpoint_policy,
+            smaller_is_better=smaller_is_better,
+            searcher_metric_name=searcher_metric_name,
+            local_training=self.local_training,
+            det_profiler=self.det_profiler,
+            records_per_epoch=records_per_epoch,
+            steps_completed=steps_completed
+        )
+
+        self.trial_controller.run()
+
+
+def initialize_distributed_backend():
+    distributed_backend = det._DistributedBackend()
+    if distributed_backend.use_horovod():
+        hvd.require_horovod_type("torch", "PyTorchTrial is in use.")
+        hvd.init()
+        return core.DistributedContext.from_horovod(horovod.hvd)
+    elif distributed_backend.use_torch():
+        if torch.cuda.is_available():
+            dist.init_process_group(backend="nccl")  # type: ignore
+        else:
+            dist.init_process_group(backend="gloo")  # type: ignore
+        return core.DistributedContext.from_torch_distributed()
+    else:
+        print(f"Backend {distributed_backend} not supported")
+
+
+@contextlib.contextmanager
+def init(hparams: Optional[Dict] = None, distributed_context: Optional[core.DistributedContext] = None):
```

Comment:
do we want to support automatically pulling hparams out of the cluster_info when running on-cluster?

That's not rhetorical, I'm not sure.

Reply:
done